### ISSUE TITLE
Implement elltool.core.AGenEllipsoid.getScalarPolar and use it everywhere instead of inline code for polar calculation

### ISSUE THREAD

**Issue by pgagarinov** [SystemAnalysisDpt-CMC-MSU/ellipsoids#28]
Deadline is 28th of October, by that time all the chagnes and the new tests should be in "master" branch
------------------------------------------------------------------------------
fSinglePolar in ellipsoid.polar and getPolar in doesContainPoly contain approximately the same code responsible for calculating polar of a single ellipsoid. Though similar the algorithms are a bit different so please implement both in a new protected elltool.core.AGenEllipsoid.getScalarPolar method. This method should have the following parameters:
1) self - the object itself (can only be scalar - check that first using isscalar and modgen.common.checkvar - see help header for checkvar and Wiki for more details).
2) isRobustMethod - if true then the algorithm from fSinglePolar  should be used, otherwise - the algorithm from doesContainPoly. 

Then replace the explicit formulas with the calls to the newly created method. Also please check all methods of ellipsoid class to see if there are some other places where this method can be used instead of explicit formulas. If there are any - make the replacement. And make sure that all the existing tests pass i.e. that you didn't break anything.

Locate existing tests for getPolar and doesContainPoly methods in elltool.core.test.mlunit package and study them. 

Then figure out a test for ellipsoids with such matrices that the not-robust method (isRobustMethod =false) used in doesContainPoly doesn't work while the robust method works. Then make doesContainPoly use the robust method also and check that this makes the test pass. Most probably the test should operate with ill-conditioned matrices of high dimensions. The test should be implemented in a separate test case class named elltool.core.test.mlunit.PolarIllCondTC

**Comment by pgagarinov** [comment]
Please

1) do not leave commented out code, even temporary. If you need to roll back some changes - use git;
2) do not commit stuff like

        disp('fSinglePolar is working');
        dbstack

You can only use it locally but please do not commit it! Please remove this code from both polar.m and getScalarPolar;

3) If you switch from fSinglePolar to getScalarPolar in polar.m then fSinglePolar should be removed.
 Please do not leave garbage code behind.

4) When checking isRobustMethod for correctness please use 'isscalar(x)&&islogical(x)', not just 'isscalar(x)'
5) make isRobustMethod optional argument (if nargin<2 isRobustMethod = false or true);
6) getScalarPolar should have a proper help header like other functions. Please use exactly the same; format (to the letter);
7) do not forget to do rebase.

**Comment by pgagarinov** [comment]
Commit "ac629b84bb1fc775647d6b294e4c30d4524330df" doesn't look like a correctly done rebase. Look at the graph structure. There are two branches merging into one. Please do the following.
1) Open log, find commit "331dc5cfb7f11b5f924450d5e77f248d94ec4c67", right click and select reset branch to this. Then select "Hard" and press Ok.
2) Do Rebase again and a) make sure you do rebase, not merge; b) if you even do "pull" please always select "launch rebase after fetch" flag.
3) After rebase please make sure that your graph structure looks like a straight line, not like two branches merging into one.
4) If everything is ok - do push after rebase. You need to select "force may discard known chagnes" flag because without this flag push won't work after rebase (this is because rebase changes history (old commits), not just adds new commits on top of old commits).

**Comment by pgagarinov** [comment]
5) you should not use names like "index" for counters - see 
Wiki for a correct naming convention.

for index = 1:numel(ellArr)
    polEllArr(index) = getScalarPolar(ellArr(index), true);

6) help header is not complete. You need to specify type and size for
each of input parameters in formal way (exactly like in Wiki).
For instance 

Not correct: 
isRobustMethod: logical: = true

Correct:

isRobustMethod: logical[1,1] - if true, then .... . Default value is ...

The same goes for self, just specifying ellipsoid type is not ok because self can be any size
(if you call getScalarPolar for an array). In you case this is a scalar so size is always [1,1]


7) modgen.common.checkmultvar('isscalar(x1)&&islogical(x2)',...
    2, self, isRobustMethod, 'errorTag', 'wrongInput');

based on this check x2 can be non-scalar? isRobustMethod is just a flag so please check if it is scalar as well.

8) if (isRobustMehtod) - round brackets "()" are not needed here

9) Please remove blank lines or replace them with comment signs "%"

10) Please avoid using simple names like "polar", "index" etc just in case. There can be a function named "polar"
somewhere and we do not want unexpected problems. Use "polarObj" instead.

**Comment by pgagarinov** [comment]
Please let me know when you fix items 1) - 10) so that I can start looking. Thanks.

**Comment by alextim27** [comment]
1)-4) I did reset and rebase, think now I did it right way. 
5)-10) I did all except right header. Will push it today. 

So, if I understand right, only thing I should do - write test, yes?

**Comment by pgagarinov** [comment]
After 1-10 are done, yes, I think the only thing remaining is the test.

**Comment by alextim27** [comment]
When I'm trying to test - have the following:
If I use non-robust method everywhere - all test passed.
If I use robust method everywhere - one test (MPTIntegrationTestCase/testRuntimeGetBoundary) doesn't work. 
So the situation is the opposit of written in issue text. What should I do in this case?

**Comment by pgagarinov** [comment]
(MPTIntegrationTestCase/testRuntimeGetBoundary)  is just a performance test so please figure out why it doesn't pass, probably some calculations take longer than before, it is normal.

In the description I asked you to develop a new test that makes the non-robust method fail: "
Then figure out a test for ellipsoids with such matrices that the not-robust method (isRobustMethod =false) used in doesContainPoly doesn't work while the robust method works." without making the robust method fail. Have you implemented such test? If not - please do so.

**Comment by alextim27** [comment]
I don't understand how I have to create test. 
I found that myGetScalarPolar function returns different results when using different methods. In particular I found that for 
ell = ellipsoid(0.01 * ones(11, 1), hilb(11)) 

shape matrix of returning ellipsoids have different like 5.6709e+10 by norm.
I can't call my function directly from test, because it's private. So if I understand right, I have to call it via doesIntersectionContain function. But I can't create such test to do this. What should I do?

**Comment by pgagarinov** [comment]
1) I can't call my function directly from test, because it's private. 
"getScalarPolar" is protected, not private. This means that you can access this method from any derivative class. This in its turn means that if you create a special class called TestEllipsoid inside a test package and define a new public method getScalarPublicPolar that calls getScalarPolar (or rename getScalarPolar into getScalarPolarInternal in AEllipsoid and call a public method in TestEllipsoid just getScalarPolar, the choise is yours). Anyways, this method will let you test getScalarPolar from AEllipsoid.

2) The idea to use the Hilbert matrix is very good. It is not only the difference in norm that matters. You  could also look at how symmetrical (if any the shape matrix is after getScalarPolar with isRobust=false). For instance, k=||A-A'|| would be a good measure of symmetry. I expect that k is lower for the robust method. If the test confirms that - the test is ready. You can just define a tolerance level kTol for k and in the test check that for the robust method k<kTol while the non-robust method k>kTol.

I hope it helps. Thanks

**Comment by alextim27** [comment]
Where TestIllipsoid should be located? In +elltool/+core ?

**Comment by pgagarinov** [comment]
Not, in a test package, let us put it into elltool.core.test.mlunit.aux

**Comment by alextim27** [comment]
So, I think I did everything, can you check?

**Comment by alextim27** [comment]
So
Tests are in PolarIllCondTC. 
I don't understand what did you mean when told about kTol, because getScalarPolar returns ellipsoid. That means shape matrix is symmetric.

**Comment by pgagarinov** [comment]
3) >I don't understand what did you mean when told about kTol, because getScalarPolar returns ellipsoid. That means shape matrix is symmetric.
You are right, I completely forgot that starting version 2.0 ET doesn't allow for slightly-nonsymmetrical matrices, in yearlier days it was
possible to create an ellispoid with [1 eps;0 1] matrix.

Anyways, testGetScalarPolar test is better than nothing because it shows that results can be very sensitive to the algorithm for ill-conditioned
matrices. But I think it is not enough because there is no ethalon matrix to compare to. 
Please come up with ill-conditioned matrix for which you can analytically calculate a polar transformation (on a piece of paper or using a
symbolic calculation software). Then you calculate this matrix in your test using 3 different ways: 

a) analytical formuala 
b) robust algorithm
c) non-robust algorithm

Then you compare a with both b and c, hopefully the result will show that b) is closer to a) than c).

4) help header in getScalarPoly is not 100% compatible with the requirement, compare your help header with the help header
in modgen.common.parseparext. It is not enough to specify type, you always need to specify size. Also please use tabs, not spaces for indentation.

5) 
        throwerror('degenerateEllipsoid',...
            'The resulting ellipsoid is not bounded');

in getScalarPolar will not work because you didn't add import modgen.common.throwerror; Please write a negative test where the resulting 
ellipsoid is not bounded.

6) Test should not display garbage information - please remote "           disp('asdfsdaf');" from your TestEllipsoid

7) Please stop putting spaces everywhere like in polarEll = getScalarPolar(ell-internalPoint, true)
In ET our policy is not to use spaces.

8) In testGetScalarPolar(self) both kTol and nDims are constants so their names should be in capital letters with underscores.

9) In testGetScalarPolar(self)  the block of code is copy-pasted, please use nested functions as I explained on last seminar 
to avoid a copy-pasting. The test should look like this:


        function self = testGetScalarPolar(self)
            COMP_TOL = 1e-2;
            %
            check(5);
            check(11);
            %
	    function check(nDims)	
       ....

**Comment by alextim27** [comment]
I have read report. 
So we have formulas for polar:
a_new = -inv(Q)(a - tetta) / (1 - K)
Q_new = inv(Q) / (1 - K) + a_new * a_new'
K = ||tetta - a||^2_inv(Q)

So if a = zeros(11, 1), Q = hilb(11) I expected (by these formulas) a_new - zeros, Q_new = inv(Q).
But robust method gives me result
norm(expShMat - shMat) = 2.9709e+11

And at the same time non-robust method
norm(expShMat - shMat) = 0

In this case nonrobust method is better, isn't it?

**Comment by pgagarinov** [comment]
The question is - how do you you calculate expShMat? If you use inv then your result just mean that non-robust method is as bad as using inv. The thing is that using "inv" to find inverse Hilber matrix is a very bad idea. This what Hilber matrix is good for - for making "inv" work terrible. 

You need to use Matlab "invhilb" function (http://www.mathworks.com/help/matlab/ref/invhilb.html). 
Of course if you already use "invhlib" then I'll investigate further but you need to commit your test so that I can have a look.

**Comment by pgagarinov** [comment]
But based based on the following calculation I have to admit that just inv is better than ell_inv for Hilbert matrix.

>> norm(invhilb(5)-ell_inv(hilb(5)))  - norm(invhilb(5)-inv(hilb(5)))

ans =

      6.82602659217196e-07

So just make sure you use invhilb in your tests and if it is confirmed that non-robust method is better - then let us just call non-robust method "robust" and vice versa and start using inv as default method.

**Comment by alextim27** [comment]
So you can see, that test I added (it uses invhilb) - Failed.

**Comment by pgagarinov** [comment]
10) Ok, so as I said, let us call inv-based method "robust" and another one - non-robust i.e. swap them in getScalarPolar. 

11) Also, please use a leading space followed by tabs in the help header.
12) Finally, please do rebase (actually you should have been doing that regularly every week at least).

**Comment by alextim27** [comment]
Ok?

**Comment by pgagarinov** [comment]
11) Not done - please check the indentation in Notepad++ or any other editor that shows tabs and spaces.

13) Please do not use ' for transposition, use .' or transpose.

**Comment by pgagarinov** [comment]
14) As far as I understand right now we use both isRobust=true and false methods in different places. Is that correct? If so -  then since you proved that just inv is getter than ell_inv please make sure that getScalarPolar is used everywhere without any input arguments and isRobust=true (inv-based) is a default method. 

I'll create the final task for you tomorrow.

**Comment by pgagarinov** [comment]
15) Copy-pasting in PolarIllCondTC: the following block of code is repeated 2 times:

            polarObj1 = self.getScalarPolarTest(ell1,true);
            polarObj2 = self.getScalarPolarTest(ell1,false);
            [~,shMat1] = double(polarObj1);
            [~,shMat2] = double(polarObj2);

I suggest you create an auxilary method in your test case like auxGetTestPolars that returns polar and shape matrices you need. Then you can use it in all your tests.

16) in the test test case Mat1 and Mat2 suffices are not correct, you can only use "Mat" suffix. Obj1 and Obj2 are also not ok, please use polar1Obj, polar2Obj, sh1Mat, sh2Mat etc. In
Q = hilb(N_DIMS);
Q is a bad name, see Wiki for variable naming convention.

**Comment by alextim27** [comment]
11), 14), 15), 16) Done

**Comment by alextim27** [comment]
And 13) also Done

**Comment by pgagarinov** [comment]
I fixed a few things on the fly in your code - please do pull with "launch rebase after fetch" but discovered a few problems:

17) TestEllipsoid should be renamed into TestPolarEllipsoid and placed into +aux sub-package. Right not it is next to the test cases which is not good.

18) Inside of TestEllipsoid you call getPolarScalar (now called getPolarScalarInternal) as a function. Please call it as a method: obj.methodName(....), not like methodName(obj)

19) Executing mlunitext.runtestcase('elltool.core.test.mlunit.PolarIllCondTC') produces 2 errors:

Error: shapeMat matrices must be symmetric and positive semi-definite, Identifier: ELLIPSOID:ELLIPSOID:wrongInput:shapeMat

it might be a result of my changes but it shouldn't... Anyway - please investigate.

20) I fixed your help header a bit - please follow the same formatting pattern in your next task. Example in the protected doesn't make any sense because such method cannot be called from outside.

**Comment by pgagarinov** [comment]
21) Just noticed that you inherited your test case from TestEllipsoid - well, this is utterly terrible thing to do, I'm not even going to describe why. just please, do not do such things any more. Instead please construct TestEllipsoid object directly. You could do that in a special private method of your test case for instance.

**Comment by alextim27** [comment]
17)-18) Done
19)  On my computer it works now
20) Ok, I understood, will follow
21) I did, but not sure did good.

**Comment by alextim27** [comment]
Oops, I did shit in lasts commits, will fix it in the nearest future

**Comment by alextim27** [comment]
Fixed. 
Can you see if I did right?

**Comment by alextim27** [comment]
As I can see, Jenkins says two errors. But when I run these tests local they passed. 
What can be the reason of this behavior?

**Comment by pgagarinov** [comment]
I think the problem is that on both my pc and build server inv of Hilber matrix results in an overflow:

K>> eig(inv(hilb(11)))

ans =

         -39206478.7085732
         -15997143.3349502
         -2174315.14275051
         -60668.8436133188
          2.66959311769106
          518.230300022676
          922791.071633742
          166099079.921795
          12032496914.7619
          1280744576812.49
           294215543744981

The maximum dimension of Hilber matrix my PC can handle without an overflow is 8.
What is your matlab version, OS and CPU model?

22) In 
test = elltool.core.test.mlunit.aux.TestPolarEllipsoid();

please replace test with something more meaningful, testEllObj for instance

obj = self.getTest();

same for obj - the problem with such names is that it is not clear what is stored in "obj", it can be anythign.
That is why names should be as self-explanatory as possible.
23) 
Some lines in PolarIllCondTC are longer than 76 symbols, please fix this.

**Comment by alextim27** [comment]
I have
Matlab 8.6.0.267246 (R2015b)
Windows 7 Intel Core i5

22)-23) done
21) ...
testGetScalarPolar results:
sh1Mat - used inv based method (now true)
sh2Mat - used previous robust method (now false)

dimension : norm(expShMat-sh1Mat) : norm(expShMat-sh2Mat) : current test result
4 :    7.2276e-10 : 5.2661e-10     ( > )  Failure
5 :  5.6985e-08 : 7.8884e-07       ( < )  Passed
6 :   8.7662e-04 : 9.9720e-04      ( < )  Passed
7 :    0.7477 : 0.6963                    ( > )  Failure
8 :  530.1504 : 223.6934              ( > )  Failure
11 :  4.3827e+11 : 7.3280e+11    ( < )  Passed

So, situation is unstable. What can we do in this case?

**Comment by pgagarinov** [comment]
Let us calculate dimVec as

DIM_VEC=2:11;isnOverflowVec=arrayfun(@(x)min(eig(inv(hilb(x)))),DIM_VEC)>0;dimVec=DIM_VEC(isnOverflowVec);

and then run your test for all dimensions from dimVec

**Comment by alextim27** [comment]
I did it in test. 
Result is the following:   

     2
   4.4409e-15
   4.4409e-15

     3
   1.3213e-12
   1.1382e-12

     4
   7.2276e-10
   5.2661e-10

     5
   5.6985e-08
   7.8884e-07

     6
   8.7662e-04
   9.9720e-04

     7
    0.7477
    0.6963

     8
  530.1504
  223.6934

    11
   4.3827e+11
   7.3280e+11

**Comment by pgagarinov** [comment]
24) You forgot to commit TestPolarEllipsoid class - please fix this
25) On my CPU (i7 5820k) I have:

>> dimVec=2:11;normDiffVec=arrayfun(@(x)(norm(invhilb(x)-ell_inv(hilb(x))) - norm(invhilb(x)-inv(hilb(x)))),dimVec).'

normDiffVec =

                         0
      1.69946440631628e-13
      5.90835231087663e-10
      6.82602659217196e-07
      8.53083103439977e-05
        -0.180588466489505
         -391.253604775645
         -529001.309870989
         -352946866.881819
         -265170786779.823

What does this command produce on your PC?

**Comment by alextim27** [comment]
On my PC
 0
         -0.00
         -0.00
          0.00
          0.00
         -0.04
       -141.99
    -352404.21
 -252304672.61
 296243002212.41

**Comment by pgagarinov** [comment]
Start working on your next task for now, I'll decide what to do by Monday.

**Comment by alextim27** [comment]
What will we do?

**Comment by pgagarinov** [comment]
Ok, here is what we do:

in testGetScalarPolar just check that normDiffVec  contains both negative and positive components (i.e. check that there is no "better" method).

Then wait for the test results - if they pass (and only if they pass) - merge to master.
When merging -  please follow this algorithm very carefully

1) Switch to master
2) Merge -> from -> origin/remotes/your branch (using local branch can dangerous because it can contain unpushed changes) 
3) Important !!! Set "no fast forward flag". Press OK
4) DO NOT PUSH JUST YET. Right click and open log. It should look like "petlya".
5) If eveything is ok - press push. 
6) Wait for the test results for updated master. If everything ok - switch to the next task which will be assigned to your very shortly.

**Comment by alextim27** [comment]
Ok. 
And finally what method will work by default? From fSinglePolar or from doesContainPoly?

**Comment by pgagarinov** [comment]
The one that uses "inv" should be default.

**Comment by pgagarinov** [comment]
No, no, no, 

check(6, true);
check(8, false);

is extremely bad way to do what we need because the on every computer sings of all components can be different. We need to check that THERE ARE both positive and negative components, no matter which:

isRobustBetterVec=....

mlunitext.assert(any(isRobustBetterVec));
mlunitext.assert(any(~isRobustBetterVec));

**Comment by pgagarinov** [comment]
26) isMethodsSimilar is a vector, where is "Vec" suffix? Check naming of variables in all your code please. And do not forget to rebase before you start merging into master because master changes constantly.

**Comment by pgagarinov** [comment]
Merge to master is done incorrectly, there should be "petelka", see how others did it (Ilya, Artem). I'm referting changes to master. Also you local branch after rebase should look without any "petelkya" i.e. it should be a straight line. Please check that first before even trying to merge again.

**Comment by pgagarinov** [comment]
Also please note that I reverted changes by Lubimov so now, when rebasing, you need to "skip" these changes in rebase dialog. If rebase says that your branch is up to date - use "force rebase". Otherwise you will put changes by Lubimov to master along with your changes. See the email I sent yo everyone for more details. Thanks.

**Comment by alextim27** [comment]
Is it ok for now?
I don't know why, but last time it was also "petlya" like this time.

**Comment by pgagarinov** [comment]
It is much better but still not 100% correct because you didn't "skip" changes made by Lubimov in master. Right now we are in a situation when after rebase you did I reverted some changes in master so master's tip has moved. But it moved not because of new changes but because of commits being reverted. So  you need to skip the commits I deleted from old master when doing rebase to the new master.

If you do rebase as usual Lubimov's changes will remain in your branch. But we need to delete them from your branch as part of rebase. This can be achieved via marking Lubimov's commits as "skip" in Rebase dialog.

**Comment by pgagarinov** [comment]
Also, a question for you - why do you ask if it ok when tests in your branch do not pass?

My point is - it cannot be ok by definition if some test doesn't pass. And you cannot merge to master. This means that you didn't check if tests pass before merging to master.

So, please re-read the procedure I sent via email and follow it to the letter. If tests do not pass- not ok and you need to fix them before even trying to merge into master (they need to pass in your branch).

**Comment by alextim27** [comment]
I did rebase and I think correctly - it looks like line, all commits from master are in, old commits by Lubimov skipped. Now Jenkins says me "Error by fetching" in log. Can you explain me, please, what I did wrong?

**Comment by pgagarinov** [comment]
Yes, your rebase looks correct.

What action did result into such an error? Please send me the complete log but I suspect that you might have some kind of hardware issue - this is not the first time you end up with a corrupted git repo. 

Anyways, if gitrepo is corrupted you always can create a fresh clone of the entire repository, after all all y our changes are pushed to the server.

Given a state of your branch you can merge but please wait for Artur Lubimov first.

**Comment by alextim27** [comment]
Once again the same problem. I just did rebase after Lubimov's updates and got error.
Log:
[build.txt](https://github.com/SystemAnalysisDpt-CMC-MSU/ellipsoids/files/43519/build.txt)

**Comment by pgagarinov** [comment]
Ok, I'm investigating this problem...

**Comment by pgagarinov** [comment]
I believe it was a temporary problem on a Jenkins build server because of some workers failing and then not being able to recover. All subsequent builds tried to use the same worker which resulted into a series of failures. I've rebooted the server and now the the build seems to have started without any problem, let us wait for the test results though just to be 100% sure.

**Comment by alextim27** [comment]
Yes, sure, waiting for results. Thanks
